DecoratedTypeName.hpp.in

```diff
@@ -0,0 +1,40 @@
+#ifndef UTILITY_DECORATEDTYPENAME_HPP
+#define UTILITY_DECORATEDTYPENAME_HPP
+
+#include "utility.hpp"
+
+namespace utility{
+
+/**
+ * @brief 
+ * Provides the typename of a base class subject to compile-time a decorator 
+ *
+ * @details
+ * Acts as a base case terminating the template recursion of the more general
+ * variadic template. 
+ */
+template< typename Base, template< typename > class Decorator >
+struct DecoratedTypeName< Base, Decorator >{
+  using value = Decorator< Base >;
+};
+
+/**
+ * @brief 
+ * Provides the typename of a base class subject to compile-time decorators
+ *
+ * @details
+ * A recursive template evaluating the type name of a base class subject to
+ * two or more compile time decorator templates. Templates applied in right to 
+ * left order. 
+ */
+template< typename Base,
+          template< typename > class Decorator,
+          template< typename > class... RemainingDecorators >
+struct DecoratedTypeName< Base, Decorator, RemainingDecorators... >{
+  using value = Decorator< typename DecoratedTypeName
+                           < Base, RemainingDecorators... >::value >;
```

Comment:
> If I use this, I get the NeutronChannelData type out of it and none of the intermediate ones?

I had a similar issue in mind when I wrote this class, but this function wasn't designed to address that particular issue. The functionality your'e looking for can be achieved using SFINAE. A jab at it:
```cpp

/* Assuming decorators define a 'core' field, whereas base implementations do not */

template< typename T, typename = int >
struct isDecorator : std:: false_type{};

template< typename T >
struct isDecorator< T, decltype((void) T::core, 0) > : std::true_type {};

template< typename T, bool  = isDecorator<T>::value >
struct UndecoratedChannelType;

template<typename T>
struct UndecoratedChannelType<T, true>{ 
  using  Name = UndecoratedChannelType< decltype(T::core) >; 
};

template<typename T>
struct UndecoratedChannelType<T, false>{
   using  Name = decltype(T);
};  
```

[Relevant reading](http://stackoverflow.com/questions/1005476/how-to-detect-whether-there-is-a-specific-member-variable-in-class)